### ISSUE TITLE
Google TeamDrives and --Fast-List not detecting all files on the remote.

### ISSUE THREAD

**Issue by Nestar47** [rclone/rclone#3114]
#### What is the problem you are having with rclone?

Files recently uploaded or transferred into a Team Drive are not detected by rclone when using the --fast-list flag. There appears to be a delay before such files appear, approx 1.5-2k files per hour in my testing. All files are properly detected when not using the flag.

#### What is your rclone version (output from `rclone version`)

v1.46.0-067-g1a019292 (Don't have access to the machine currently)

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Windows 2008 SBS 64-bit

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive / Team Drives

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

rclone sync

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

Will provide next week

**Comment by Nestar47** [comment]
Based on https://forum.rclone.org/t/fast-list-overwriting-and-endless-sync-with-gsuite-drive/8718/18

**Comment by Nestar47** [comment]
rclone v1.46.0-067-g1a019292-fix-3032-suffix-keep-extension-beta
- os/arch: windows/amd64
- go version: go1.11

Zipping -vv from both a standard run and the first part of a --fast-list (stopped as soon as it starts uploading files). Will email shortly.

**Comment by ncw** [comment]
That is strange... The main difference between the non fast list and the fast list is that we use batch operations to list more directories at once.

Can you try playing with this constant to see if it makes a difference?  Try making it smaller

https://github.com/ncw/rclone/blob/f5795db6d2fb6b321f0568c418ea86698fa11490/backend/drive/drive.go#L1489

@B4dM4n any ideas?

**Comment by Nestar47** [comment]
Ya, I see that non-fast is requesting each folder by itself. The fast-list requests are just getting a blank response back. See line 9277 on the fastlist log for the first one. 9245 above it was successful.

```
c
{"files":[]}
0
```

The folders all appear to be listed, just the item responses are being returned blank.

I unfortunately am not setup to build. I'll test a modified version if you have it though.

**Comment by ncw** [comment]
Yes that looks like the problem. This query

```
2019/04/15 08:37:47 DEBUG : GET /drive/v3/files?alt=json&corpora=teamDrive&fields=files%28id%2Cname%2Csize%2Cmd5Checksum%2Ctrashed%2CmodifiedTime%2CcreatedTime%2CmimeType%2Cparents%2CwebViewLink%29%2CnextPageToken&includeTeamDriveItems=true&pageSize=1000&prettyPrint=false&q=trashed%3Dfalse+and+%28%XXX%27+in+parents+or+%27XXX%27+in+parents+or+%27XXX%27+in+parents+or+...+in+parents%29&supportsTeamDrives=true&teamDriveId=XXX HTTP/1.1
```

Returned

```
{"files":[]}
```

Which is surely wrong.

So this **is** a bug in google drive I'd say.

However, can we work around it...?

Here is an rclone with `--drive-fast-list-grouping` - this defaults to 50 - can you try experimenting with it to see if you can it to work at some value?  I'd try 25,10,5,2,1

https://beta.rclone.org/branch/v1.47.0-006-gad7078c3-fix-3114-teamdrive-listr-beta/ (uploaded in 15-30 mins)

You could also try `rclone size` just to count the files.  This will use `--fast-list` by default now - if you don't want it to, use `--disable ListR`

**Comment by Nestar47** [comment]
Ran with Non, Fast, 25, 10, 5 and 1.
1 behaves exactly like Non Fast-List. The rest are all missing files (the number varies slightly)

`c:\rclone\test>rclone size Drive1_19k:Test --disable ListR
Total objects: 109345
Total size: 3.851 GBytes (4135425516 Bytes)

c:\rclone\test>rclone size Drive1_19k:Test
Total objects: 945
Total size: 26.572 MBytes (27863240 Bytes)

c:\rclone\test>rclone size Drive1_19k:Test --drive-fast-list-grouping 25
Total objects: 2835
Total size: 77.405 MBytes (81165497 Bytes)

c:\rclone\test>rclone size Drive1_19k:Test --drive-fast-list-grouping 10
Total objects: 3980
Total size: 109.123 MBytes (114423575 Bytes)

c:\rclone\test>rclone size Drive1_19k:Test --drive-fast-list-grouping 5
Total objects: 3558
Total size: 99.308 MBytes (104131739 Bytes)

c:\rclone\test>rclone size Drive1_19k:Test --drive-fast-list-grouping 1
Total objects: 109345
Total size: 3.851 GBytes (4135425516 Bytes)
`

Looking closer into the responses

**Comment by Nestar47** [comment]
Unfortunately, it appears that the response to an actual empty folder is the exact same as the blank response we're getting from the batch requests

Eg, in the nonfastlist.log I sent previously

Rclone initially learns about it's existence in the response at 9672.
line 11296 is the request for the items inside the blank folder
Response comes back at 11599


With that in mind I see a few options 

1. Assume a blank response to a --fast-list request is invalid and queue up those items as individual requests as well.

Pros:
Should always find the files
Transparent to the user
Cons:
Potentially more API calls then just running without --fast-list (At least off the bat, until google drive catches up)

2. Warn/Error when blank responses are received Eg. "RClone received a blank response from a Google Drive request, This can sometimes indicate that files have not yet been indexed on the remote side. Consider running without --fast-list"

Pros:
Won't waste API calls
Notifies the user of a potential issue

Cons:
There are legitimate blank folder situations that will trigger the warning.
Situations where it returns some files but not all of them would not be detectable.
Might not be visible if the user is running a headless setup.

3. Disable --fast-list on team drives until google can fix their end.

Pros:
Should always work
Cons:
Slower and more API calls
Probably a bit excessive.

**Comment by ncw** [comment]
Option 1 sounds feasible...

Option 2 sounds workable too, since all the directories you were querying would have to be empty to produce a false positive.

3. Also a good plan... I couldn't find an existing issue about this though - can you? ttps://issuetracker.google.com/

**Comment by ncw** [comment]
This is the best I could come up with: https://issuetracker.google.com/issues/75001471

**Comment by Nestar47** [comment]
Ya, that looks to be the same symptoms. That one only has a single item so it populates rather quickly, I'm surprised it was even detectable.
Having 100k+ like I've got is amplifying it considerably.

**Comment by Nestar47** [comment]
> Option 2 sounds workable too, since all the directories you were querying would have to be empty to produce a false positive.

I'd be more concerned about false negatives in that situation. I've only seen examples of returning all data and returning no data, I'm guessing it's also possible to return only some of the data. Finding an example of this might be tricky.

> This is the best I could come up with: https://issuetracker.google.com/issues/75001471

His work-around of adding a 3 second delay before the query for a single file does seem to match my approximate 1500-2000 files per hour measurement as well. If the "bug" on google's end truly is a throttle or account limitation then there may not ever be a fix.

**Comment by Nestar47** [comment]
Just got the email for changes to Drive API, including some adjustments to team drives. 

https://cloud.google.com/blog/products/application-development/upcoming-changes-to-the-google-drive-api-and-google-picker-api

https://developers.google.com/drive/api/v3/reference/files/list

**Comment by ncw** [comment]
I had a go at making a workaround here - can you have a test?

https://beta.rclone.org/branch/v1.52.0-004-g96c6137c-fix-4289-listr-shared-beta/ (uploaded in 15-30 mins)

    drive: fix missing items when listing using --fast-list / ListR
    
    This is caused by a bug in Google drive where, in some circumstances
    querying for "(A in parents) or (B in parents)" returns nothing
    whereas querying for "A in parents" and "B in parents" separately
    works fine.
    
    This has been reported here:
    
    https://issuetracker.google.com/issues/149522397
    
    This workaround detects this condition by seeing if a listing for more
    than one directory at once returns nothing.
    
    If it does then it retries each one individually.
    
    This can potentially have a false positive if the user has multiple
    empty directories which are queried at once. The consequence of this
    will be that ListR is disabled for a while until the directories are
    found to be actually empty in which case it will be re-enabled.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52.1
